src/overlays/actors/ovl_En_Tp/z_en_tp.c

```diff
@@ -58,43 +79,699 @@ static ColliderJntSphInit D_80B22AB4 = {
+    EnTp_SetupAction(this, EnTp_Head_Wait);
+}
+
+/**
+ * Awaken and rise from the ground when Player is closer than 200
+ */
+void EnTp_Head_Wait(EnTp* this, GlobalContext* globalCtx) {
+    Player* player = PLAYER;
+    s16 yaw;
+
+    this->unk_15C--;
+
+    if (this->actor.xzDistToPlayer < 200.0f) {
+        if (this->collider.base.atFlags & AT_HIT) {
+            this->collider.base.atFlags &= ~AT_HIT;
+            if (&player->actor == this->collider.base.at) {
+                this->timer = 0;
+            }
+        }
+
+        if (this->timer != 0) {
+            this->timer--;
+
+            Math_SmoothStepToS(&this->actor.shape.rot.x, 0, 1, 500, 0);
+            Math_SmoothStepToS(&this->actor.world.rot.y, this->actor.yawTowardsPlayer, 1, 1500, 0);
+
+            yaw = Math_Vec3f_Yaw(&this->actor.home.pos, &player->actor.world.pos) + 0x4000;
+            Math_SmoothStepToF(&this->actor.world.pos.y, this->actor.home.pos.y + 30.0f, 0.3f, 1.0f, 0.3f);
+            this->actor.world.pos.x = this->actor.home.pos.x +
+                                      (Math_SinS(2000 * this->unk_15C) * (Math_SinS(yaw) * this->horizontalVariation));
+            this->actor.world.pos.z = this->actor.home.pos.z +
+                                      (Math_SinS(2000 * this->unk_15C) * (Math_CosS(yaw) * this->horizontalVariation));
+        } else {
+            this->actor.shape.rot.x = 0;
+            this->unk_150 = 1;
+            EnTp_Head_SetupTakeOff(this);
+        }
+    } else {
+        Math_SmoothStepToS(&this->actor.shape.rot.x, -0x4000, 1, 500, 0);
+
+        if (Math_SmoothStepToF(&this->actor.world.pos.y, this->actor.home.pos.y, 0.3f, 1.5f, 0.3f) == 0.0f) {
+            this->timer = 60;
+        } else {
+            yaw = Math_Vec3f_Yaw(&this->actor.home.pos, &player->actor.world.pos);
+            this->actor.world.pos.x =
+                this->actor.home.pos.x + (Math_SinS(2000 * this->unk_15C) * (Math_SinS(yaw) * 6.0f));
+            this->actor.world.pos.z =
+                this->actor.home.pos.z + (Math_SinS(2000 * this->unk_15C) * (Math_CosS(yaw) * 6.0f));
+        }
+    }
+
+    this->actor.shape.rot.y = this->actor.world.rot.y;
+
+    if (this->actor.world.pos.y != this->actor.home.pos.y) {
+        Audio_PlaySoundGeneral(NA_SE_EN_TAIL_FLY - SFX_FLAG, &this->actor.projectedPos, 4, &D_801333E0, &D_801333E0,
+                               &D_801333E8);
+    }
+}
+
+void EnTp_Head_SetupBurrowReturnHome(EnTp* this) {
+    this->actionIndex = TAILPASARAN_ACTION_HEAD_BURROWRETURNHOME;
+    this->timer = 0;
+    EnTp_SetupAction(this, EnTp_Head_BurrowReturnHome);
+}
+
+void EnTp_Head_BurrowReturnHome(EnTp* this, GlobalContext* globalCtx) {
+    static Vec3f bubbleAccel = { 0.0f, -0.5f, 0.0f };
+    static Color_RGBA8 bubblePrimColor = { 255, 255, 255, 255 };
+    static Color_RGBA8 bubbleEnvColor = { 150, 150, 150, 0 };
+    Vec3f bubbleVelocity;
+    Vec3f bubblePos;
+    s32 closeToFloor;
+    EnTp* now;
```

Comment:
```suggestion
    EnTp* child;
```
Or something similar.

Reply:
I'm not sure about this, since it iterates over the generations of children

Reply:
I don't think `now` is a good name for it. `child` isn't a good alternative either imo.

Reply:
Yeah, happy for suggestion of alternatives. It's used to iterate over the tail segments, which are linked using parent-child, ending at the head.